Unified Restriction

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

For the hackathon. This PR attempts to reduce the number of kernel calls in prolongation and restriction by coalescing them into meshblock packs.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented. (none)
- [x] Adds a test for any bugs fixed. Adds tests for new features. (no new features/bugs)
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] (@lanl.gov employees) Update copyright on changed files

jonahm-LANL committed:
removed unused function

jonahm-LANL committed:
rename some things

jonahm-LANL committed:
move restriction into a team policy design

jonahm-LANL committed:
formatting

Yurlungur committed:
Merge branch 'develop' into jmm/unified-prolongation

jonahm-LANL committed:
linting and changelog

jonahm-LANL committed:
it compiles for GPU

jonahm-LANL committed:
partial commit

jonahm-LANL committed:
fix stupid index error

jonahm-LANL committed:
segfault somewhere

jonahm-LANL committed:
fix stupid typo

jonahm-LANL committed:
remove dead code

jonahm-LANL committed:
formatting with clang-format 8

jlippuner commented on src/bvals/boundary_conditions.cpp:
```diff
@@ -34,8 +34,9 @@ TaskStatus ProlongateBoundaries(std::shared_ptr<MeshBlockData<Real>> &rc) {
   // physical variables in:
-  // - step 2, ApplyPhysicalBoundariesOnCoarseLevel(): calls to W(U) and user BoundaryFunc
-  // - step 3, ProlongateGhostCells(): calls to calculate bcc and U(W)
+  // - 1: Apply the physical boundaries on the course level
```
```suggestion
  // - 1: Apply the physical boundaries on the coarse level
```

jlippuner commented on src/bvals/bvals.hpp:
```diff
@@ -145,12 +145,9 @@ class BoundaryValues : public BoundaryBase, // public BoundaryPhysics,
 
   // ProlongateBoundaries() wraps the following S/AMR-operations (within nneighbor loop):
```
```suggestion
  // ProlongateBoundaries() wraps the following S/AMR-operations (within neighbor loop):
```

jlippuner commented on src/mesh/mesh_refinement.hpp:
```diff
@@ -77,6 +77,8 @@ class MeshRefinement {
   int AddToRefinement(ParArrayND<Real> pvar_cc, ParArrayND<Real> pcoarse_cc);
   int AddToRefinement(FaceField *pvar_fc, FaceField *pcoarse_fc);
 
+  Coordinates_t GetCoarseCoords() const { return coarse_coords; }
```
Maybe return a const ref?
```suggestion
  const Coordinates_t &GetCoarseCoords() const { return coarse_coords; }
```

Yurlungur replied:
I don't think that works--I need `coarse_coords` to go to the GPU so I don't want it to be a reference.

jlippuner commented on src/mesh/refinement_cc_in_one.hpp:
```diff
@@ -0,0 +1,36 @@
+// Copyright(C) 2020 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020-2021. Triad National Security, LLC. All rights
```
```suggestion
// (C) (or copyright) 2021. Triad National Security, LLC. All rights
```

jlippuner commented on src/mesh/refinement_cc_in_one.hpp:
```diff
@@ -0,0 +1,36 @@
+} // namespace cell_centered_refinement
+} // namespace parthenon
+
+#endif // MESH_REFINEMENT_CC_IN_ONE_HPP_
```
Newline at the end?

Yurlungur replied:
fixed

jlippuner commented on src/kokkos_abstraction.hpp:
```diff
@@ -591,6 +597,65 @@ inline void par_for_outer(OuterLoopPatternTeams, const std::string &name,
+  const int NjNi = Nj * Ni;
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, NjNi), [&](const int &idx) {
+    int j = idx / Ni + jl;
+    int i = idx % Ni + il;
```
Not sure if it will make a difference, but modulo could be slow on GPU. But integer multiplication is also not the fastest...

Yurlungur replied:
Yeah that's worth looking at. I didn't think too hard about this.

jlippuner reviewed (approved):
Looks good to me. Just some small non-blocking comments.

Yurlungur committed:
Update src/bvals/boundary_conditions.cpp

Co-authored-by: Jonas Lippuner <jlippuner@lanl.gov>

Yurlungur committed:
Update src/bvals/bvals.hpp

Co-authored-by: Jonas Lippuner <jlippuner@lanl.gov>

Yurlungur committed:
Update src/mesh/refinement_cc_in_one.hpp

Co-authored-by: Jonas Lippuner <jlippuner@lanl.gov>

Yurlungur committed:
Update refinement_cc_in_one.hpp

Yurlungur commented:
Huh... The advection performance test is failing on the GPU on the CI. Not on my local V100 though.

forrestglines reviewed (approved):
Looks good, but I can't see why `nvprof` segfaults in the CI.

pgrete commented:
I'll dig into the segfault as I have direct access to the machine.
Also I'd like to get some performance numbers of this PR before we merge.
I plan to write a new performance test based on the example we used for the hackathon and ideally integrate with the #395  machinery first so that we have that automated for various scenarios.
Also, did you test performance on Phoebus with these changes?

Yurlungur commented:
> I'll dig into the segfault as I have direct access to the machine.

Thanks! That's useful!

> Also I'd like to get some performance numbers of this PR before we merge.
> I plan to write a new performance test based on the example we used for the hackathon and ideally integrate with the #395 machinery first so that we have that automated for various scenarios.

That sounds good!

> Also, did you test performance on Phoebus with these changes?

I haven't had the chance to do that yet, but will do so this week.

pgrete commented:
@Yurlungur can I push directly to this branch?
I was able to reproduce the segfault that shows up in CI on a different machine and it likely needs some fixes around ` 1 0x0000000000465926 parthenon::cell_centered_bvars::ResetSendBufferBoundaryInfo()` that I'm happy to hunt down.

Yurlungur commented:
Please do push directly, @pgrete ! Thanks for hunting this down!

Yurlungur commented:
@pgrete I found what may be a related bug. There were a few calls to boundary data that should only be called for multilevel, but which were being called for single level. I will fix.

jonahm-LANL committed:
fix bug where AMR stuff is exposed when AMR is off

jonahm-LANL committed:
Merge branch 'jmm/unified-prolongation' of github.com:lanl/parthenon into jmm/unified-prolongation

Yurlungur commented:
> Also, did you test performance on Phoebus with these changes?

I have some preliminary tests of this now... We got a huge performance boost from hooking in the old pack-in-one machinery. But the hackathon branch appears to be somewhat worse. I see about a 30% improvement in the re-meshing time, but per-cycle, the code is a little slower.

I need to dig in to this a bit more. So far,, all I've done is look at the per-cycle timing output.

jonahm-LANL committed:
remove restriction from the prolongate_boundaries call and allow boundary comms to do the restriction instead

jonahm-LANL committed:
fix copyrights

jonahm-LANL committed:
formatting

pgrete committed:
Merge branch 'develop' into jmm/unified-prolongation

jonahm-LANL committed:
Re-enable restriction in ProlongateBoundaries, but do it the clever way. This breaks things downstream.

Yurlungur commented:
Some updates:
- The reason the code slowed down per cycle was that the old restriction routines were still called in `ProlongateBoundaries`. At first I thought this call was unnecessary. However it appears that restriction *is* necessary there still, as some coarse ghost zones need to be filled for prolongation that are not filled during boundary communication.
- I have updated the `ProlongateBoundaries` call. This is, unfortunately, API breaking. Here's what I did:
  - Remove `RestrictBoundaries` from the `ProlongateBoundaries` call. Instead `RestrictBoundaries` is its own task, on which `ProlongateBoundaries` must depend
  - Do a buffer filling/pack-in-one kind of thing for restriction, that can be enabled or disabled. The per meshblock task is still available.

Unfortunately, there seems to be some kind of issue. When making this change, I introduced a 1e-9 error in the `output_hdf5` test, which I'm having trouble tracking down. 1e-9 is a very small error, so maybe it's nothing to worry about and we should just loosen the testing constraints or make a new gold file. But it also seems very unusual and it's bothering me.

I would welcome so extra eyes on this PR if they're available. @pgrete, you've already been digging in so maybe you'd see something. All eyes would be welcome though.

Yurlungur committed:
Merge branch 'develop' into jmm/unified-prolongation

Yurlungur commented:
@pgrete some extra eyes on this to understand why tests are failing would be very welcome

pgrete committed:
Merge branch 'develop' into jmm/unified-prolongation

pgrete committed:
Add restriction to MeshInit

Yurlungur commented:
Thanks for the catch, @pgrete ! Running the full tests now. Let's see what it looks like.

Yurlungur commented:
All tests now pass. I'm going to do a little profiling and then I'll pull the trigger.

Yurlungur committed:
Merge branch 'develop' into jmm/unified-prolongation

pgrete commented:
FYI using #534  on this branch and compare to currently develop results in 
![image](https://user-images.githubusercontent.com/22685568/120015780-0575ff80-bfe4-11eb-9974-7f672cf5b263.png)

pgrete commented:
Also for reference, reducing the block size to 16 in the amr_performance test (with the following mesh evol):
```
cycle=0 time=0.0000000000000000e+00 dt=8.7890624999999991e-04 zone-cycles/wsec_step=0.00e+00 wsec_total=3.32e-02 wsec_step=1.18e+01 zone-cycles/wsec=0.00e+00 wsec_AMR=0.00e+00
---------------------- Current Mesh structure ----------------------
Root grid = 8 x 8 x 8 MeshBlocks
Total number of MeshBlocks = 1632
Number of physical refinement levels = 3
Number of logical  refinement levels = 6
  Physical level = 0 (logical level = 3): 448 MeshBlocks, cost = 448
  Physical level = 1 (logical level = 4): 448 MeshBlocks, cost = 448
  Physical level = 2 (logical level = 5): 480 MeshBlocks, cost = 480
  Physical level = 3 (logical level = 6): 256 MeshBlocks, cost = 256
--------------------------------------------------------------------
cycle=1 time=8.7890624999999991e-04 dt=8.7890624999999991e-04 zone-cycles/wsec_step=2.16e+06 wsec_total=3.12e+00 wsec_step=3.09e+00 zone-cycles/wsec=2.16e+06 wsec_AMR=6.85e-05
cycle=2 time=1.7578124999999998e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.10e+06 wsec_total=4.44e+00 wsec_step=1.31e+00 zone-cycles/wsec=5.10e+06 wsec_AMR=7.16e-05
cycle=3 time=2.6367187499999997e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=4.92e+06 wsec_total=5.79e+00 wsec_step=1.36e+00 zone-cycles/wsec=4.92e+06 wsec_AMR=3.21e-04
cycle=4 time=3.5156249999999997e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.10e+06 wsec_total=7.11e+00 wsec_step=1.31e+00 zone-cycles/wsec=5.10e+06 wsec_AMR=6.52e-05
cycle=5 time=4.3945312500000000e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.12e+06 wsec_total=8.41e+00 wsec_step=1.31e+00 zone-cycles/wsec=5.12e+06 wsec_AMR=6.20e-05
cycle=6 time=5.2734375000000003e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.05e+06 wsec_total=9.73e+00 wsec_step=1.32e+00 zone-cycles/wsec=5.05e+06 wsec_AMR=6.94e-05
cycle=7 time=6.1523437500000007e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.08e+06 wsec_total=1.27e+01 wsec_step=1.31e+00 zone-cycles/wsec=2.22e+06 wsec_AMR=1.70e+00
-------------- New Mesh structure after (de)refinement -------------
Root grid = 8 x 8 x 8 MeshBlocks
Total number of MeshBlocks = 1716
Number of physical refinement levels = 3
Number of logical  refinement levels = 6
  Physical level = 0 (logical level = 3): 448 MeshBlocks, cost = 448
  Physical level = 1 (logical level = 4): 448 MeshBlocks, cost = 448
  Physical level = 2 (logical level = 5): 468 MeshBlocks, cost = 468
  Physical level = 3 (logical level = 6): 352 MeshBlocks, cost = 352
--------------------------------------------------------------------
cycle=8 time=7.0312500000000010e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=4.30e+06 wsec_total=1.44e+01 wsec_step=1.63e+00 zone-cycles/wsec=4.30e+06 wsec_AMR=7.55e-05
cycle=9 time=7.9101562500000014e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=5.01e+06 wsec_total=1.58e+01 wsec_step=1.40e+00 zone-cycles/wsec=5.01e+06 wsec_AMR=1.67e-04
cycle=10 time=8.7890625000000017e-03 dt=8.7890624999999991e-04 zone-cycles/wsec_step=4.89e+06 wsec_total=1.72e+01 wsec_step=1.44e+00 zone-cycles/wsec=4.89e+06 wsec_AMR=1.26e-04
```
results in

![image](https://user-images.githubusercontent.com/22685568/120018155-0d836e80-bfe7-11eb-9cb5-95420ecf39b4.png)

pgrete commented on src/bvals/cc/bvals_cc_in_one.cpp:
```diff
@@ -292,6 +278,18 @@ void ResetSendBufferBoundaryInfo(MeshData<Real> *md, size_t buffers_used) {
+
+  // Get coarse and fine bounds. Same for all blocks.
+  auto &rc = md->GetBlockData(0);
+  auto pmb = rc->GetBlockPointer();
```
Potentially add scope or rename `pmb` as `pmb` is also used in the following for loop.

Yurlungur replied:
This one is essentially a throw-away, and it's fine that it's shadowed in scope. We only use it to get the index shapes, etc. But I agree, it could be hidden in `{}`.

Yurlungur committed:
Merge branch 'develop' into jmm/unified-prolongation

Yurlungur commented:
Nice! THanks for computing that, @pgrete ! Overall that's quite a satisfying performance improvement! WIth these in place, I'm going to go ahead and merge when tests pass.

pgrete reviewed (commented):
I have one non-blocking comment/question:
As far as I read the code there's now a new `BufferCache_t restrict_buffers_` objects, but the machinery also works on the `send_buffers_` object (as it's called in ResetSendBufferBoundaryInfo` with `cell_centered_refinement::Restrict(boundary_info, cellbounds, c_cellbounds);`
Where's the difference between these two caches (with respect to the restriction) and could we get away with just one (in a future PR)?

Yurlungur commented:
> As far as I read the code there's now a new `BufferCache_t restrict_buffers_` objects, but the machinery also works on the `send_buffers_` object (as it's called in ResetSendBufferBoundaryInfo`with`cell_centered_refinement::Restrict(boundary_info, cellbounds, c_cellbounds);`
> Where's the difference between these two caches (with respect to the restriction) and could we get away with just one (in a future PR)?

Right. I initially tried making these two caches the same. Unfortunately, the restriction targets are different. Consider 3 meshblocks on a coarse fine-boundary:
```
C | F
  | F
```
The `send_buffers_` object restricts to send to the coarse block. The fine blocks need the data from the coarse block to fill their ghost zones.  However, they *also* need data on their coarse buffers from *each other*  for the full prolongation stencil. This is where the `restrict_buffers_` cache comes in. This was also the source of a bug I struggled with initially---as the current code is written, restriction needs to be performed *twice*: once to send to coarse blocks and once to fill the coarse buffers on fine blocks in preparation for prolongation.

I'm not sure whether or not these two cache objects could be unified. The design which was easier for me to reason about was the one I went with---but I think it might be possible to simplify things.
